azurelinuxagent/pa/provision/cloudinit.py

```diff
@@ -132,3 +126,29 @@ def wait_for_ssh_host_key(self, max_retry=1800, sleep_time=1):
         raise ProvisionError("Giving up, ssh host key was not found at {0} "
                              "after {1}s".format(path,
                                                  max_retry * sleep_time))
+
+def cloud_init_is_enabled():
+    """
+    Determine whether or not cloud-init is enabled.
+
+    Args:
+        None
+
+    Returns:
+        bool - True if cloud-init is enabled, False if otherwise.
+    """
+
+    try:
+        systemctl_output = subprocess.check_output([
+            'systemctl',
+            'is-enabled',
+            'cloud-init-local.service'
+        ], stderr=subprocess.STDOUT).decode('utf-8').replace('\n', '')
```

Comment:
This will produce an exception if the service is missing.

I'm also not sure using `systemctl` is a particular good idea.

Reply:
An exception is ok. This is wrapped in a try/except so that any exception case (such as a missing service) will just result in a falsy return. As for using `systemctl`, it is the best option forward from what I've seen. I originally tried using dbus but there is no guarantee that the system dbus is up and running at this point, whereas systemd has it's own internal dbus for communication (which systemctl uses) that we can rely on at this boot stage. Let me know if you have any other hesitations for using systemctl otherwise.

Reply:
I am not certain that calling systemctl from within a code path that runs as part of a systemd unit will always produce reliable results. I should but I don't know for certain nor have I found any documentation thereof.

This also makes the assumption that the generator (which is safe) has run, and properly detected Azure (there have been generator detection issues). Last but not least it assumes that all images run the generator that is part of cloud-init. There are other way to run `cloud-init-local.service`. While we can argue that that's not the "standard way" when/if someone has a "brilliant idea" and this code breaks I did want to point out at the implicit dependency that is being created by this check.

I am not advocating for a change, just making a comment.